src/types.ts

```diff
@@ -1423,6 +1424,26 @@ export class ZodArray<
       }
     }
 
+    if (def.uniq) {
+      (() => {
+        for (let i = 0; i < ctx.data.length; i++) {
+          const a = ctx.data[i];
+
+          for (let j = i + 1; j < ctx.data.length; j++) {
+            const b = ctx.data[j];
+
+            if (a === b) {
+              addIssueToContext(ctx, {
+                code: ZodIssueCode.duplicate_element,
+              });
+              status.dirty();
+              return;
+            }
+          }
+        }
+      })();
```

Comment:
I'm not sure why you used an IIFE with a double for loop here.
I think a really easy and quick approach would be using `Set` in something like:

```suggestion
if (new Set(ctx.data).size < ctx.data.length) {
  addIssueToContext(ctx, {
    code: ZodIssueCode.duplicate_element,
  });
  status.dirty();
}
```

Reply:
I used an IIFE with double loop for execution complexity optimisation, this way I'm able to stop the processing as soon as I find a duplicates avoiding to check the rest of the array if it's not necessary. Using A Set would mean to check duplicates for all elements of the array then check the lenght of the Set to get the result. The code would be simpler but the execution would be worse.

Reply:
Did you test performance of double loop over a Set so you say it's more optimized?

Reply:
By the way in case of `unique array` schema may be prefer to use `z.set` instead of array?

Reply:
Effectively I didn't tested it before committing it. I just made a few tests and they show that the double loop as heavily better performance on arrays with duplicates but worst performance on arrays without duplicates. I found an other implementation with object and simple loop, here is my test code: 
``` ts
import _ from 'lodash';

function checkDuplicateDoubleLoop(list: any[]) {
  return (() => {
    for (let i = 0; i < list.length; i++) {
      const a = list[i];

      for (let j = 0; j < i; j++) {
        if (a === list[j]) {
          return false;
        }
      }
    }
    return true;
  })();
}

function checkDuplicateSet(list: any[]) {
  return new Set(list).size >= list.length;
}

function checkDuplicateObject(list: any[]) {
  return (() => {
    const obj: any = {};
    for (const elem of list) {
      if (obj[elem]) {
        return false;
      } else {
        obj[elem] = true;
      }
    }
    return true;
  })();
}

function checkDeepDuplicateDoubleLoop(list: any[]) {
  return (() => {
    for (let i = 0; i < list.length; i++) {
      const a = list[i];

      for (let j = 0; j < i; j++) {
        if (_.isEqual(a, list[j])) {
          return false;
        }
      }
    }
    return true;
  })();
}

function checkDeepDuplicateSet(list: any[]) {
  return new Set(list).size >= list.length;
}

(() => {
  const list = [];

  for (let i = 0; i < 10000000; i++) {
    list.push(i);
  }

  // commented because too long

  // console.time('noDupDoubleLoop');
  // console.log('noDupDoubleLoop: ', checkDuplicateDoubleLoop(list));
  // console.timeEnd('noDupDoubleLoop');

  console.log('noDupDoubleLoop: Too long');

  console.time('noDupSet');
  console.log('noDupSet: ', checkDuplicateSet(list));
  console.timeEnd('noDupSet');

  console.time('noDupObject');
  console.log('noDupObject: ', checkDuplicateObject(list));
  console.timeEnd('noDupObject');

  list.unshift(0);

  console.time('dupDoubleLoop');
  console.log('dupDoubleLoop: ', checkDuplicateDoubleLoop(list));
  console.timeEnd('dupDoubleLoop');

  console.time('dupSet');
  console.log('dupSet: ', checkDuplicateSet(list));
  console.timeEnd('dupSet');

  console.time('dupObject');
  console.log('dupObject: ', checkDuplicateObject(list));
  console.timeEnd('dupObject');
})();

(() => {
  const list = [];

  for (let i = 0; i < 10000; i++) {
    list.push({ i });
  }

  console.time('noDeepDupDoubleLoop');
  console.log('noDeepDupDoubleLoop: ', checkDeepDuplicateDoubleLoop(list));
  console.timeEnd('noDeepDupDoubleLoop');

  console.time('noDeepDupSet');
  console.log('noDeepDupSet: ', checkDuplicateSet(list));
  console.timeEnd('noDeepDupSet');

  console.time('noDeepDupObject');
  console.log('noDeepDupObject: ', checkDuplicateObject(list));
  console.timeEnd('noDeepDupObject');

  list.unshift({ i: 0 });

  console.time('deepDupDoubleLoop');
  console.log('deepDupDoubleLoop: ', checkDeepDuplicateDoubleLoop(list));
  console.timeEnd('deepDupDoubleLoop');

  console.time('deepDupSet');
  console.log('deepDupSet: ', checkDuplicateSet(list));
  console.timeEnd('deepDupSet');

  console.time('deepDupObject');
  console.log('deepDupObject: ', checkDuplicateObject(list));
  console.timeEnd('deepDupObject');
})();
```

and results : 

``` bash
noDupDoubleLoop: Too long
noDupSet:  true
noDupSet: 2.184s
noDupObject:  true
noDupObject: 742.271ms
dupDoubleLoop:  false
dupDoubleLoop: 0.232ms
dupSet:  false
dupSet: 2.179s
dupObject:  false
dupObject: 0.099ms
noDeepDupDoubleLoop:  true
noDeepDupDoubleLoop: 31.074s
noDeepDupSet:  true
noDeepDupSet: 1.935ms
noDeepDupObject:  false // expect true
noDeepDupObject: 0.114ms
deepDupDoubleLoop:  false
deepDupDoubleLoop: 0.396ms
deepDupSet:  true // expect false
deepDupSet: 1.172ms
deepDupObject:  false
deepDupObject: 0.222ms
```

I will update my PR to implement the object solution.
You can also notice that I tested Set and object solution for deep object unicity check and for that check only the Double loop solution is working. I will so implement the deepUniq function with the double loop solution (I made the deep check using lodash for going faster for testing but for final implementation I will make it in an hand made function to avoid adding lodash in Zod dependencies just for that utility.

PS: trying to go further in testing I also noticed that on big arrays with more than 10 000 000 items you can reach Set maximun size limit. The Set solution would then be not suitable for a production reliable librairy like Zod  (even if I do not recommend using the uniq check on such arrays because of processing time expect for exceptional use cases)